generic/tests/unattended_install.py

```diff
@@ -11,4 +27,57 @@ def run(test, params, env):
     :param params: Dictionary with the test parameters.
     :param env: Dictionary with test environment.
     """
-    unattended_install.run(test, params, env)
+    ret = ThreadError()
+    # use vm name to differentiate among single/multivm
+    vms = params.get("vms").split()
+    nr_vms = len(vms)
+    multivm = nr_vms > 1
+    trigger_time = int(params.get("install_trigger_time", 0))
+    random_trigger = params.get("random_trigger", "no") == "yes"
+    install_timeout = int(params.get("install_timeout", 200))
+    thread_timeout = (1 + 0.05 * nr_vms) * install_timeout
+    ret_lock = threading.RLock()
+    install_threads = []
+
+    def thread_func(vm):
+        """
+        Thread Method to trigger the unattended installation
+
+        :param vm: VM name
+        """
+        try:
+            vm_params = params.object_params(vm)
+            vm_params["main_vm"] = vm
+            unattended_install.run(test, vm_params, env)
+        except Exception as info:
+            logging.error(info)
+            ret_lock.acquire()
+            ret.error()
+            ret_lock.release()
+
+    if multivm:
+        if not params.get("master_images_clone"):
+            test.cancel("provide the param `master_images_clone` to clone"
+                        "images for vms")
+    for vm in vms:
+        if multivm:
+            thread = threading.Thread(target=thread_func, args=(vm,))
+            install_threads.append(thread)
+        else:
+            thread_func(vm,)
+
+    if multivm:
+        for thread in install_threads:
+            if trigger_time:
+                sleep_time = trigger_time
+                if random_trigger:
+                    sleep_time = random.randint(0, trigger_time)
+                time.sleep(sleep_time)
+            thread.start()
+    for thread in install_threads:
+        thread.join(thread_timeout)
```

Comment:
By this, you will increase the per-thread `install_timeout` by few more times: suppose that there are two vms to be installed, and the `install_timeout` is 200s, and assumed that `thread_timeout` is equal to `install_timeout`, if both the installation of the each vm run out of time, for the first vm it would wait 200s and that is correct, but for the second vm it would wait another 200s (in total 400s) which is incorrect.

Reply:
no if the thread is started at the same time then all the vms would get same 200s to timeout. if second vm takes another 200 s then there is no point in using threads it is just linear execution. And it is better to use `install_timeout`  in `thread.join()` rather forming the math function.

Reply:
No, they do start at the same time but you just joined them one by one. Here is an example which is the equivalent.
```python
import time
from threading import Thread

DURATION = 10
TIMEOUT = 3
WORKERS = 5

SCHEDULE = {}

def workload(wid, duration):
    global SCHEDULE
    SCHEDULE[wid] = time.time()
    time.sleep(duration)

workers = [Thread(target=workload, args=(i, DURATION))
           for i in range(WORKERS)]
for w in workers:
    w.start()

print("timeout is %ds" % TIMEOUT)
for i, w in enumerate(workers):
    w.join(TIMEOUT)
    live_time = time.time() - SCHEDULE[i]
    print("worker #%d run out of time after %ds" % (i, live_time))
```
And the result is:
```
$ python a.py
timeout is 3s
worker #0 run out of time after 3s
worker #1 run out of time after 6s
worker #2 run out of time after 9s
worker #3 run out of time after 10s
worker #4 run out of time after 10s
```
and you can see only the first worker get the correct deadline.

Reply:
Hi @luckyh, Thank you for your explaining with example code. so I tried it with below changes and all the worker get uniform timeout,
```python
for i, w in enumerate(workers):
    # include the time worker takes to sleep
    TIMEOUT += DURATION
    w.join(TIMEOUT)
    live_time = time.time() - SCHEDULE[i]
    print("worker #%d run out of time after %ds" % (i, live_time))
```
we can get all workers to get uniform timeout,
**Output:**
```
$ python thread.py 
timeout is 3s
worker #0 run out of time after 10s
worker #1 run out of time after 10s
worker #2 run out of time after 10s
worker #3 run out of time after 10s
worker #4 run out of time after 10s
```

Reply:
so if the when the last 5th worker thread starts, it could get timeout accumulated with the sleep time of first 4 worker thread, does this justifies ? or is there anything I miss.